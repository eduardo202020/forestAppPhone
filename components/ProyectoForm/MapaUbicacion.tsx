import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import MapView, { Marker } from 'react-native-maps';

interface MapaUbicacionProps {
  selectedLocation: { latitude: number; longitude: number } | null;
  mapRef?: React.RefObject<MapView>;
  onMapPress?: (e: any) => void;
  editable?: boolean;
  error?: string;
}

const MapaUbicacion: React.FC<MapaUbicacionProps> = ({
  selectedLocation,
  mapRef,
  onMapPress,
  editable = true,
  error,
}) => (
  <>
    <Text style={styles.label}>Ubicación</Text>
    <MapView
      ref={mapRef}
      style={styles.map}
      initialRegion={{
        latitude: selectedLocation?.latitude || -12.0464,
        longitude: selectedLocation?.longitude || -77.0428,
        latitudeDelta: 0.0922,
        longitudeDelta: 0.0421,
      }}
      onPress={editable ? onMapPress : undefined}
    >
      {selectedLocation && <Marker coordinate={selectedLocation} />}
    </MapView>
    <Text style={styles.ubicacionText}>
      {selectedLocation
        ? `Ubicación seleccionada: ${selectedLocation.latitude.toFixed(5)}, ${selectedLocation.longitude.toFixed(5)}`
        : 'Toca el mapa para seleccionar una ubicación'}
    </Text>
    {error && <Text style={styles.errorText}>{error}</Text>}
  </>
);

const styles = StyleSheet.create({
  map: {
    width: '100%',
    height: 180,
    borderRadius: 10,
    marginBottom: 10,
  },
  ubicacionText: {
    fontSize: 12,
    color: '#888',
    marginBottom: 10,
  },
  label: {
    fontWeight: 'bold',
    marginBottom: 4,
    marginTop: 8,
  },
  errorText: {
    color: 'red',
    marginBottom: 8,
    fontSize: 13,
  },
});

export default MapaUbicacion;
