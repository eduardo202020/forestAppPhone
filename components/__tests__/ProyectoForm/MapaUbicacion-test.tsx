import React from 'react';
import renderer from 'react-test-renderer';
import MapView from 'react-native-maps';
import MapaUbicacion from '../../ProyectoForm/MapaUbicacion';

describe('MapaUbicacion', () => {
  it('se renderiza sin errores', () => {
    const tree = renderer
      .create(
        <MapaUbicacion
          selectedLocation={{ latitude: -12.0464, longitude: -77.0428 }}
          editable
          error={undefined}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('muestra el error si existe', () => {
    const tree = renderer
      .create(
        <MapaUbicacion
          selectedLocation={null}
          editable
          error="Campo requerido"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('llama a onMapPress cuando el usuario presiona el mapa', () => {
    const onMapPress = jest.fn();
    const testRenderer = renderer.create(
      <MapaUbicacion selectedLocation={null} editable onMapPress={onMapPress} />
    );
    // Simula el evento onPress del MapView
    const map = testRenderer.root.findByType(MapView);
    map.props.onPress({
      nativeEvent: { coordinate: { latitude: 1, longitude: 2 } },
    });
    expect(onMapPress).toHaveBeenCalled();
  });
});
