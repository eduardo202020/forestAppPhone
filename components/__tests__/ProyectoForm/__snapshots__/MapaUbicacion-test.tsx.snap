// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapaUbicacion muestra el error si existe 1`] = `
[
  <Text
    style={
      {
        "fontWeight": "bold",
        "marginBottom": 4,
        "marginTop": 8,
      }
    }
  >
    Ubicaci贸n
  </Text>,
  <AIRMap
    initialRegion={
      {
        "latitude": -12.0464,
        "latitudeDelta": 0.0922,
        "longitude": -77.0428,
        "longitudeDelta": 0.0421,
      }
    }
    onChange={[Function]}
    onMapReady={[Function]}
    region={null}
    style={
      {
        "borderRadius": 10,
        "height": 180,
        "marginBottom": 10,
        "width": "100%",
      }
    }
  />,
  <Text
    style={
      {
        "color": "#888",
        "fontSize": 12,
        "marginBottom": 10,
      }
    }
  >
    Toca el mapa para seleccionar una ubicaci贸n
  </Text>,
  <Text
    style={
      {
        "color": "red",
        "fontSize": 13,
        "marginBottom": 8,
      }
    }
  >
    Campo requerido
  </Text>,
]
`;

exports[`MapaUbicacion se renderiza sin errores 1`] = `
[
  <Text
    style={
      {
        "fontWeight": "bold",
        "marginBottom": 4,
        "marginTop": 8,
      }
    }
  >
    Ubicaci贸n
  </Text>,
  <AIRMap
    initialRegion={
      {
        "latitude": -12.0464,
        "latitudeDelta": 0.0922,
        "longitude": -77.0428,
        "longitudeDelta": 0.0421,
      }
    }
    onChange={[Function]}
    onMapReady={[Function]}
    region={null}
    style={
      {
        "borderRadius": 10,
        "height": 180,
        "marginBottom": 10,
        "width": "100%",
      }
    }
  >
    <AIRMapMarker
      coordinate={
        {
          "latitude": -12.0464,
          "longitude": -77.0428,
        }
      }
      onPress={[Function]}
      style={
        [
          {
            "backgroundColor": "transparent",
            "position": "absolute",
          },
          undefined,
        ]
      }
    />
  </AIRMap>,
  <Text
    style={
      {
        "color": "#888",
        "fontSize": 12,
        "marginBottom": 10,
      }
    }
  >
    Ubicaci贸n seleccionada: -12.04640, -77.04280
  </Text>,
]
`;
